espnet/nets/pytorch_backend/e2e_asr_mulenc.py

```diff
@@ -0,0 +1,601 @@
+        self.init_like_chainer()
+
+        # options for beam search
+        if args.report_cer or args.report_wer:
+            recog_args = {'beam_size': args.beam_size, 'penalty': args.penalty,
+                          'ctc_weight': args.ctc_weight, 'maxlenratio': args.maxlenratio,
+                          'minlenratio': args.minlenratio, 'lm_weight': args.lm_weight,
+                          'rnnlm': args.rnnlm, 'nbest': args.nbest,
+                          'space': args.sym_space, 'blank': args.sym_blank,
+                          'tgt_lang': False, 'ctc_weights_dec': self.weights_ctc_dec}
+
+            self.recog_args = argparse.Namespace(**recog_args)
+            self.report_cer = args.report_cer
+            self.report_wer = args.report_wer
+        else:
+            self.report_cer = False
+            self.report_wer = False
+        self.rnnlm = None
+
+        self.logzero = -10000000000.0
+        self.loss = None
+        self.acc = None
+
+    def init_like_chainer(self):
+        """Initialize weight like chainer.
+
+        chainer basically uses LeCun way: W ~ Normal(0, fan_in ** -0.5), b = 0
+        pytorch basically uses W, b ~ Uniform(-fan_in**-0.5, fan_in**-0.5)
+
+        however, there are two exceptions as far as I know.
+        - EmbedID.W ~ Normal(0, 1)
+        - LSTM.upward.b[forget_gate_range] = 1 (but not used in NStepLSTM)
+        """
+        def lecun_normal_init_parameters(module):
+            for p in module.parameters():
+                data = p.data
+                if data.dim() == 1:
+                    # bias
+                    data.zero_()
+                elif data.dim() == 2:
+                    # linear weight
+                    n = data.size(1)
+                    stdv = 1. / math.sqrt(n)
+                    data.normal_(0, stdv)
+                elif data.dim() in (3, 4):
+                    # conv weight
+                    n = data.size(1)
+                    for k in data.size()[2:]:
+                        n *= k
+                    stdv = 1. / math.sqrt(n)
+                    data.normal_(0, stdv)
+                else:
+                    raise NotImplementedError
+
+        def set_forget_bias_to_one(bias):
+            n = bias.size(0)
+            start, end = n // 4, n // 2
+            bias.data[start:end].fill_(1.)
+
+        lecun_normal_init_parameters(self)
+        # exceptions
+        # embed weight ~ Normal(0, 1)
+        self.dec.embed.weight.data.normal_(0, 1)
+        # forget-bias = 1.0
+        # https://discuss.pytorch.org/t/set-forget-gate-bias-of-lstm/1745
+        for l in six.moves.range(len(self.dec.decoder)):
+            set_forget_bias_to_one(self.dec.decoder[l].bias_ih)
+
+    def forward(self, xs_pad_list, ilens_list, ys_pad):
+        """E2E forward.
+
+        :param List xs_pad_list: list of batch (torch.Tensor) of padded input sequences
+                                [(B, Tmax_1, idim), (B, Tmax_2, idim),..]
```

Comment:
In this case, is it correctly working with `DataParallel`?
In my understanding, `DataParallel` split the data according to the first axis, so it seems to cause the error.
If it does not work with `DataParallel`, it is better to add `raise NotImplementedError` in `asr.py` when `num_enc > 1` and add `TODO` comment.

Also, adding the multi gpu case test is great.

Reply:
Thanks a lot. It is not gonna work with DataParallel for now. I have made a TODO in the asr.py and made changes accordingly.

Reply:
I think data-parallel support can be delayed. do you agree @kan-bayashi?

Reply:
I agree.

Reply:
Thanks @kan-bayashi @ShigekiKarita for the reviews and comments. Is there any other suggestions for this PR?